File: ShootR/ShootR/Server/Abilities/AbilityHandlers/AbilityHandler.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;

namespace ShootR
{
    public class AbilityHandler
    {
        private Dictionary<string, Ability> _abilityList;

        public AbilityHandler(params Ability[] abilityList)
        {
            _abilityList = new Dictionary<string, Ability>();

            foreach (Ability ability in abilityList)
            {
                _abilityList.Add(ability.Name, ability);
            }
        }

        public Dictionary<string, Ability> Abilities()
        {
            return _abilityList;
        }

        public Ability Ability(string abilityName)
        {
            return _abilityList[abilityName];
        }

        public void AddAbility(Ability ability)
        {
            _abilityList.Add(ability.Name, ability);
        }

        public void Activate(string abilityName)
        {
            if (_abilityList.ContainsKey(abilityName))
            {
                _abilityList[abilityName].Activate();
            }
        }

        public void Deactivate(string abilityName)
        {
            if (_abilityList.ContainsKey(abilityName))
            {
                _abilityList[abilityName].Deactivate();
            }
        }

        public void Update(DateTime utcNow)
        {
            foreach (Ability ability in _abilityList.Values)
            {
                ability.Update(utcNow);
            }
        }
    }
}

File: ShootR/ShootR/Client/Ships/Abilities/AbilityHandlers/AbilityHandler.ts
/// <reference path="../../../../Scripts/endgate-0.2.0.d.ts" />
/// <reference path="../Ability.ts" />

module ShootR {

    export class AbilityHandler {
        private _abilityList: { [s: string]: Ability; } = {};

        constructor(aList: Array<Ability>) {
            for (var i = aList.length - 1; i >= 0; i--) {
                this._abilityList[aList[i].Name] = aList[i];
            }
        }

        public Abilities(): { [s: string]: Ability; } {
            return this._abilityList;
        }

        public Ability(abilityName: string): Ability {
            return this._abilityList[abilityName];
        }

        public AddAbility(ability: Ability): void {
            this._abilityList[ability.Name] = ability;
        }

        public Activate(abilityName: string): boolean {
            if (this._abilityList[abilityName]) {
                this._abilityList[abilityName].Activate();
                return true;
            }
            return false;
        }

        public Deactivate(abilityName: string): boolean {
            if (this._abilityList[abilityName] && this._abilityList[abilityName].Active) {
                this._abilityList[abilityName].Deactivate();
                return true;
            }
            return false;
        }

        public UpdateAbilities(aList: any): void {
            for (var abilityName in aList) {
                var dataActive = aList[abilityName],
                    myActive = this._abilityList[abilityName].Active;

                if (dataActive && !myActive) {
                    this.Activate(abilityName);
                }
                else if (!dataActive && myActive) {
                    this.Deactivate(abilityName);
                }
            }
        }

        public Update(gameTime: eg.GameTime): void {
            for (var key in this._abilityList) {
                this._abilityList[key].Update(gameTime);
            }
        }
    }

}